servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.server.authn;
+
+import io.quarkus.security.AuthenticationFailedException;
+import io.quarkus.security.identity.IdentityProviderManager;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.identity.request.AnonymousAuthenticationRequest;
+import io.quarkus.security.identity.request.AuthenticationRequest;
+import io.quarkus.vertx.http.runtime.security.ChallengeData;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticationMechanism;
+import io.quarkus.vertx.http.runtime.security.HttpCredentialTransport;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.projectnessie.quarkus.config.QuarkusNessieAuthenticationConfig;
+
+/**
+ * HTTP authentication mechanism implementation to disallow all unauthorized accesses, except for
+ * the explicitly configured paths and static resources.
+ */
+@ApplicationScoped
+public class NessieAuthMechanism implements HttpAuthenticationMechanism {
+
+  /**
+   * {@code HttpAuthenticationMechanism}s with a higher priority are "asked" first, let
+   * `NessieAuthMechanism` be called quite late, basically as the last one.
+   */
+  public static final int PRIORITY = 0;
+
+  private final boolean enabled;
+  private final AuthenticationFailedException unauthorized;
+  private final Set<String> anonymousPaths;
+  private final Set<String> anonymousPathPrefixes;
+
+  @Inject
+  public NessieAuthMechanism(QuarkusNessieAuthenticationConfig config) {
+    this.enabled = config.enabled();
+    this.unauthorized = new AuthenticationFailedException("Missing or unrecognized credentials");
+    Set<String> paths = new HashSet<>();
+    config.anonymousPaths().ifPresent(paths::addAll);
+    config.anonymousPathPrefixes().ifPresent(paths::addAll);
+    this.anonymousPaths = paths;
+    this.anonymousPathPrefixes =
+        config
+            .anonymousPathPrefixes()
+            .map(prefixes -> prefixes.stream().map(p -> p + '/').collect(Collectors.toSet()))
+            .orElse(Set.of());
+  }
+
+  boolean checkPath(String path) {
+    // Allow certain preconfigured paths (e.g. health checks) to be serviced without
+    // authentication.
+    return anonymousPaths.contains(path)
+        || anonymousPathPrefixes.stream().anyMatch(path::startsWith);
+  }
+
+  @Override
+  public int getPriority() {
+    return PRIORITY;
+  }
+
+  @Override
+  public Uni<SecurityIdentity> authenticate(
+      RoutingContext context, IdentityProviderManager identityProviderManager) {
+    if (enabled) {
+      var path = context.request().path();
+      if (checkPath(path)) {
+        return Uni.createFrom().nullItem();
```

Comment:
Should we use an `AnonymousAuthenticationRequest` is this case rather than return null?

Reply:
See https://github.com/projectnessie/nessie/pull/9845#discussion_r1827412980

Reply:
I'm not sure that comment applies. Returning `null` means this 'mechanism" did not handle the request (so the next mechanism will have a try), but the intention appears to be that we want the request to be handles specifically as "anonymous".

Reply:
I think the new Nessie code relies on Quarkus to attempt creating an anonymous identity automatically if none of the authentication mechanisms returned an identity... IMHO, this is not intuitive. Would you mind adding a comment about that? Otherwise, the connection between `NessieAuthMechanism` and `NessieAnonymousIdentityProvider` is rather obscure.

Reply:
Also, Using `0` priority does not guarantee that there's no other auth mechanism after Nessie :shrug:

Reply:
Why not get a `IdentityProviderManager` injected and do an explicit `.authenticate(AnonymousAuthenticationRequest.INSTANCE)` request?

Reply:
> Why not do an explicit `.authenticate(AnonymousAuthenticationRequest.INSTANCE)` request?

https://github.com/projectnessie/nessie/pull/9845#discussion_r1827412980 : "...because `IdentityProvider#authenticate` requires a 2nd `AuthenticationRequestContext`, which we don't have there."

Reply:
> Also, Using `0` priority does not guarantee that there's no other auth mechanism after Nessie 🤷

"Multiple mechanisms are sorted in descending order..."

Reply:
Priority `-1` is allowed :shrug:

Reply:
Re: anonymous identity, I mean this:
```
diff --git a/servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java b/servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java
index b4a726c9d..9c3a7e9ae 100644
--- a/servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java
+++ b/servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java
@@ -84,7 +84,7 @@ public class NessieAuthMechanism implements HttpAuthenticationMechanism {
     if (enabled) {
       var path = context.request().path();
       if (checkPath(path)) {
-        return Uni.createFrom().nullItem();
+        return identityProviderManager.authenticate(AnonymousAuthenticationRequest.INSTANCE);
       }
 
       if (!path.contains("..")) {
```